### DISCUSSION TITLE
Absolute include paths not working properly after 10.7.0 upgrade

### DISCUSSION THREAD

**Discussion by grivard** [cypress-io/cypress#23713] [Questions and Help]
I'm having a problem with my Cypress upgrade (9.6.0 to 10.7.0) and looking for some help or advice.  After upgrading, absolute paths seem to be partially broken.  I say partially because they work fine when my config file is at the project root, but I get errors when it's nested in a folder, where I need it to be (we have several config files for different test environments).

Our test configuration (from project root) was like this for 9.6.0:
 cypress.config.json <-- see note* below
 config\*.json <-- .json config files
 integration\various_hierarchies\*.js <-- these are the spec files, which use absolute paths to include library files
 support\*.js <-- these are the said library files, which also use absolute paths to include other library files

*Normally, we don't have a cypress.config at project root.  I moved one of the other files from config and renamed it cypress.config.json so that Cypress would modify it for me on upgrade.  It did and the subset of tests I tried using that configuration ran as expected.

 cypress.config.js <-- file created by Cypress from cypress.config.json, worked out of the box.  Plugins were handled like this:
`    setupNodeEvents(on, config) {
      return require('./elsag/plugins')(on, config)
    },
`
I did experiment with setting up the plugins directly and bypassing plugins.index.js, but since we have multiple config files and this works, it seems like a better solution so I left it as is.

Except that I wanted it to be an absolute path, so I removed the leading . - which caused Cypress to fail with the following error:
![image](https://user-images.githubusercontent.com/58826572/189005678-183f3bfa-bd6d-47e5-b002-f735b9cb5256.png)

I restored the relative path to cypress.config.js, copied the file to config/ and renamed it, and adjusted its relative path for the plugins location.

With that setup, my spec files work, including plugins.  That is, unless they include a library using absolute paths (which most of them do of course) in which case I'll get this error: 
![image](https://user-images.githubusercontent.com/58826572/189005795-395978b9-ddbf-4b5f-8f53-cef46624bc24.png)

It goes on to list other files that truly do not exist, but /elsag/support/audit-library.js most certainly does exist, assuming that / signifies project root.

The project root is 'C:\Workspace\Testing\Functional Testing\' and 'C:\Workspace\Testing\Functional Testing\elsag\support\audit-library.js' does exist.  And it will be found and will work, as long as I reference it using a relative path.

Note that in the error above, Cypress seems to be looking for the file in the integration folder, where it does not exist.  'C:\Workspace\Testing\Functional Testing\elsag\integration\elsag\support\audit-library.js' does not exist.

Spec files that include libraries using absolute paths work just fine, as long as my config file is at the root.  It doesn't need to be named cypress.config.js, as long as it lives at project root.

Does anybody understand what's going on here?  Thanks for taking the time to read this in any case.

**Comment by lmiller1990** [comment]
I tried this on linux and it's working fine:

```sh
$ cat /tmp/blah.js          
                                                                                                                                        
export default (on, config) => { 
  console.log('ok')
  return config 
}

$ cat cypress.config.ts       
                                                                                                                                      
import { defineConfig } from 'cypress'

export default defineConfig({
  fixturesFolder: false,
  e2e: {
    setupNodeEvents(on, config) {
      const blah = require('/tmp/blah.js')
      return blah.default // need to do this for es modules
    },
    baseUrl: 'http://localhost:8888',
  },
})
```

I wonder if it's windows specific? Or I am misunderstanding the instructions?

Another way you can make an absolute path (and be sure it's really absolute) is 

```js
setupNodeEvents (on, config) {
  const absPluginsPath = path.join(config.projectRoot, "cypress", "plugins")
}
```

`config.projectRoot` is the absolute path to your project.

I don't see any documentation on how Cypress handled absolute paths w.r.t `pluginsFile` in Cypress 9 - it looks like, from your post, in Cypress 9 `/` for a pluginsFile would be absolute, but Cypress 10 is interpreting it differently?

**Reply by grivard** [reply]
Thanks for your reply, Lachlan!  I don't fully understand your linux config example, but it seems like you're adding a 'default' method to every include file, then calling it in the config for every include file.  If that's right, it would be easier for me to just grit my teeth and change all my specs and includes to relative paths (or move my config files to the project root).

To clarify the problem a bit (I hope):

I have config files like this (leading /=project root):
/config/*.js

I have include files like this:
/elsag/support/*.js

I have spec files like this:
/elsag/integration/**/*.js (where ** may mean any number of sub-folders)

The config file must reference /elsag/plugins using a relative path:
    setupNodeEvents(on, config) {
      return require('./../elsag/plugins')(on, config)
    },

No big deal here.  I would prefer an absolute path, but we don't have all that many config files and they'll all be in the same directory, so an easy fix.

The bigger problem is that pretty much all of our spec files begin something like this:
import '/elsag/support/audit-library.js'
import { KendoGrid } from '/elsag/support/grid-classes.js'

If my config file is at project root, this works (as it always did in Cypress < 10, even with my config files in /config/*.json).

But in 10 if my config file is in /config/*.js, the include files cannot be found unless I change them to relative paths.
import './../elsag/support/audit-library.js'
import { KendoGrid } from './../elsag/support/grid-classes.js'
(and any files those files reference must also be changed to relative paths)

I've confirmed that Cypress.config('projectRoot') is the same regardless of where the config file is.  

The error message suggests that it's looking for the library in the integration folder rather than project root.
Module not found: Error: Can't resolve '/elsag/support/audit-library.js' in 'C:\Workspace\Testing\Functional Testing\elsag\integration'

But that looks like a red herring, since if I put the config file at the root and misspell the name of a library, the resulting error message also shows the integration folder.
Module not found: Error: Can't resolve '/elsag/support/audit-libraryx.js' in 'C:\Workspace\Testing\Functional Testing\elsag\integration'

I'm at a loss at this point.  I am thinking about ways I might make this work with one configuration file at the root and using variables to set needed properties, or maybe by using a technique like the one described at https://docs.cypress.io/api/plugins/configuration-api#Switch-between-multiple-configuration-files.

But I'd really like to understand the problem first if at all possible.  I've seen posts that say absolute paths need to be configured in JavaScript, but 1) I never had to configure them in the first place as best I can recall; and 2) I can't find a good page that tells me _how_ to configure them for JavaScript (TypeScript yes, but I'm not using TypeScript and don't know anything about it).  Also, if they really got unconfigured by the upgrade, then a) the next upgrade may break my fix assuming I can find one; and b) that might justify this being considered a bug rather than a configuration problem.

Does that help to clarify the problem?  Any other ideas?

**Reply by lmiller1990** [reply]
There Is a huge amount of information here and I'm having trouble digesting it all - is there a more concise way to state the actual problem?

**Reply by lmiller1990** [reply]
>  but it seems like you're adding a 'default' 

This was just for combining CJS (module.exports) and ESM(import/export). If you import an ESM file into CJS, you have to do `default` to actually the values.

**Reply by lmiller1990** [reply]
If you can provide a minimal reproduction of your exact problem, happy to make some suggestions. I am not sure if we regressed or not, it seems like we might have made a change to something that was working but undocumented (or not covered by our tests).

As an aside, using absolute paths seems like it's not a great idea in general - it won't work cross platform, and potentially won't work across machines with different usernames. relative imports seems like a good way to handle this, or the path.join(projectRoot, ...). strategy.

**Reply by grivard** [reply]
Thanks for your responses.  I may be able to come up with a concise example, and I'm pretty sure I can give you a more concise explanation.  Sorry for not doing so before (but hopefully the long version may be useful as a reference).

When my config file is at project root, import statements that use absolute paths work.
When my config files is nested in a folder, import statements that use absolute paths do not work.  References must be changed to relative paths in this case.

As far as the rationale for using absolute vs. relative paths, all machines where tests will be developed (that's 2 guys' desktops) or run (that's 2 automated test VMs) use the same path, which is connected to the same repository.  Cypress is a significant part of the infrastructure, but there are other tools too.  There are many benefits to having a set path structure for all of them.

I'll see if I can make it happen with as few files as possible.  Again, thanks for responding.

**Reply by lmiller1990** [reply]
Ok I got it now, I think. So.

> When my config file is at project root, import statements that use absolute paths work.

## Scenario 1 - works

/dev
/dev/project/package.json
/tmp/blah.js
/dev/project/cypress.config.js

You execute `npx cypress open` in `/dev/project`. Command is `npx cypress open`.

```js
// /dev/project/cypress.config.js

module.exports = {
  e2e: {
    setupNodeEvents(on, config) {
      const blah = require('/tmp/blah.js')    
    }
  }
}
```

## Scenario 2 - does not work

/dev
/dev/project/package.json
/dev/project/packages/nested-1/package.json
/dev/project/packages/nested-2/cypress.config.js
/tmp/blah.js

You execute `npx cypress open --project packages/nested-1` in `/dev/project`.

```js
// /dev/project/packages/nested-2/cypress.config.js
module.exports = {
  e2e: {
    setupNodeEvents(on, config) {
      const blah = require('/tmp/blah.js')    
    }
  }
}
```

Is this correct? Is the key difference you execute the `npx cypress` command in a directory that is **not** where the `cypress.config.js` lives?

**Reply by grivard** [reply]
Not exactly, but close.  And yes, I think the key difference is that I execute npx cypress in a directory that is not where the config file is; but in my case I execute from a parent of the config folder; the config is not outside of the project root folder.

Both scenarios start from C:\Workspace\Testing\Functional Testing\

### Scenario 1 - works

package.json
cypress.config.js <-- config file at project root
dev/plugins/index.js
dev/support/library.js
dev/spec-folder/spec.js

```
// cypress.config.js

module.exports = defineConfig({
  e2e: {
    setupNodeEvents(on, config) {
      return require('./dev/plugins')(on, config) // always have to use relative path here
    }
  }
})
```

```
// dev/spec-folder/spec.js

import '/dev/support/library.js  // this library is found
```
I start Cypress from a batch file that does this:

```
cd "C:\Workspace\Testing\Functional Testing\"
npx cypress open -C cypress.config.js
```
### Scenario 2 - doesn't work

package.json
config/cypress.config.js <-- config file nested in folder
dev/plugins/index.js
dev/support/library.js
dev/spec-folder/spec.js

```
// cypress.config.js

module.exports = defineConfig({
  e2e: {
    setupNodeEvents(on, config) {
      return require('./../dev/plugins')(on, config) // always have to use relative path here
    }
  }
})
```

```
// dev/spec-folder/spec.js

import '/dev/support/library.js  // this library is not found
import '../dev/support/library.js  // this library would be found
```

I start Cypress from a batch file that does this:

```
cd "C:\Workspace\Testing\Functional Testing\"
npx cypress open -C config\cypress.config.js
```

The only real difference between scenarios 1 & 2 is that I moved the config file into a config folder.  All other adjustments are simply to support that.

**Reply by lmiller1990** [reply]
Ok, got it. I think we've got a minimal reproduction here... sounds like the issue is "absolute imports behave differently when using a non-default `cypress.config` path".

If you want to put together a minimal reproduction someone can clone with instructions on reproducing it, and create an issue, someone might be able to look into it. It sounds like a bug - whether it can be fixed without breaking other workflows is yet to be seen, but it'd be nice to at least file a formal bug report with a reproducible example to document it, and eventually get prioritised.

Are you willing to make a PR if I offer some guidance about where to look in the cod base?

**Reply by grivard** [reply]
Sure.  I'll try to put together a simple project that illustrates this issue.  If it is in fact a bug, I'm surprised it hasn't been reported already; but I guess to your earlier point absolute paths may not be as common as I'd thought they'd be.

Whatever guidance you can provide is most appreciated.  I did submit a bug report a few years back, so I was able to follow the instructions at that time; as I recall I had to make a repository.  I'll review that over the weekend, but in the meantime welcome your suggestions.

**Reply by lmiller1990** [reply]
Let's start with an issue, we can go from there.

I think absolute paths is probably not common for the reason it doesn't work on every machine, unless they've got the exactly same project hierarchy.

**Reply by grivard** [reply]
I had some unrelated troubles creating a simplest case illustration of this, but I'll get back to it and hopefully post an issue sometime this week.  I installed 10.8.0 btw (was hopeful for a mystery fix after I saw 23426), but issue still occurs.

**Reply by lmiller1990** [reply]
Ok, just tag me when you make the issue and reference this discussion. Thanks!

**Reply by grivard** [reply]
I have to give up my attempts to make a simplest case illustration of this problem.  I have a functional workaround, so I can't justify putting any more time into trying to identify exactly where it falls apart, but Cypress doesn't seem to me to be behaving consistently with regard to absolute paths.  Hopefully I can provide enough information for you to determine whether there are really bugs here and if so how to reproduce them.

In pre-10 Cypress, import paths worked consistently like this: I could put lines like the following into any spec or support file, no matter where it was nested in my hierarchy, and it would find the referenced file.
`import '/elsag/support/some-library.js'`
`import { ClassName } from '/elsag/support/some-class.js'`

These lines would refer to C:\Workspace\Testing\Functional Testing\elsag\support\*.js, which would be found.  Project root was (and still is) C:\Workspace\Testing\Functional Testing.

In Cypress 10, as long as I keep my config files at project root, the above continues to work.  But if I put my config files in C:\Workspace\Testing\Functional Testing\config (where they were in Cypress <10 and where I'd prefer them), the include files referred to in the above manner are not found.

It seems to me that this has to be a bug.  Seems like the behavior should be the same no matter where the config file lives.

While trying to reproduce this problem using a simpler project. I found that it would work if I specified the full path:
`import '/workspace/testing/functional testing/elsag/support/some-library.js'`

I went back to the original project and tried updating a couple of my files using this longer path.  When I do that, my spec file is able to find the support file.  However, if the support file also has references to other libraries, those references don't work from the library itself.  So if some-library.js contains a line like the following, another-library.js won't be found:
`import '/workspace/testing/functional testing/elsag/support/another-library.js'`

Along the way, I've also found what I think is another bug having to do with case-sensitivity of specPattern when included in a command line.  Although the 4 machines that have Cypress installed here all have the same hierarchy, it turns out that my machine was C:\Workspace and the other three were C:\workspace.  This is the same path in Windows, of course, but I found that when I reference the specPattern in a command line, case matters.  I had to change all references to "C:\Workspace\Testing\Functional Testing\elsag\integration\*.js" to "C:\workspace\Testing\Functional Testing\elsag\integration\*.js".

Seems to me this also has to be a bug, since case is irrelevant in Windows path references.  

As I said above, I can't justify putting any more time into this.  I hope there is enough info in this ticket, and particularly this comment, for you to reproduce the bugs if there are any.

Our installation is like this (for all users):
![image](https://user-images.githubusercontent.com/58826572/193691242-4c24af45-67fc-45f5-9301-4326c9714af6.png)

If you need anything else from me I'll be happy to provide it.  If you need me to try anything, I'll be happy to do that too.  Any documentation you can point me to that describes how absolute paths are _supposed_ to work in Cypress 10 would likely be helpful - especially if my workaround is exploiting an undocumented 'feature' that could blow up on me with any upgrade.

Thank you so much for your help on this.  I hope you can get to the bottom of it, because I think there are real issues here, even if I _am_ doing something incorrectly.

**Reply by lmiller1990** [reply]
Hi, thanks for this. Sorry I did not respond - I was out of the office (and off grid!) for all of October. I am back now.

Regarding the case sensitive issue, we use [globby](https://www.npmjs.com/package/globby), which might be case sensitive. I wonder if we should just do `toLowerCase` on everything to simplify this.

For your actual issue of inconsistency around absolute paths - I'd like to look into it a bit more... it looks like there is something fishy going on. I don't think I can pick this up immediately, but I'll keep this open and at the back of my mind, and if I'm in that area of the code base in the future, poke around and see if I can reproduce it. Thanks!

**Reply by grivard** [reply]
No problem.  Hope you enjoyed your (I assume) vacation, and thanks for picking up the ball on this.

Something that might help would be an understanding of what's _supposed_ to happen with absolute paths.  Sounds like a silly question until we consider that absolute paths in this context are not _really_ absolute.

Before Cypress 10, we could begin a path with a / anywhere and it would be "absolutely relative" to the project root.  (I'm referring here to the use of import, include, and require within any .js file.)

As long as we store our config files at project root, that continues to work everywhere in Cypress 10 except in the config file itself when referencing plugins.  No big deal, we just made it relative for now.

When we move our config files to a nested folder ({project root}/config/config-file.js), these 'absolutely relative from project root' paths no longer work.

In trying to find a simplest case, I discovered that I could reproduce this issue using "absolute" paths that are relative to project root.  However, when I tried a path that was "absolute" relative to the drive, it worked.  I tried that in our project, but it didn't work - possibly because the files referred to also had "absolute" references to project root.  It's been a few weeks so I don't recall how far I took it, I just reached a point where I decided to stop chasing the squirrel.

If you can tell me what "absolute" paths are supposed to be relative to, I might have better luck getting you repro steps.  All I can say now is that handling of absolute paths is inconsistent depending on the location of the config file, and it's possible there may be other factors since I recall some inconsistent results when I tried to narrow it down.

As far as case-sensitivity goes, yes it would seem that using toLowerCase would eliminate case-based discrepancies.  A non-issue for us now (we have 2 live users and 2 shared VMs, so it was easy to work around).

Thanks again for your persistance!

**Reply by lmiller1990** [reply]
> Something that might help would be an understanding of what's supposed to happen with absolute paths

I think the concept of an absolute path should really be what it is everywhere else in your operating system - absolute from the root. In linux-y OSes, it would be `/`, and on windows, it would be `C:`.

WDYT? If we can move forward with this simple definition, it will be easy to make everything consistent. If it means breaking changes, we will either need to do is quickly before v11 (going out soon) or it'll need to wait for v12.

**Reply by grivard** [reply]
<blockquote>WDYT? If we can move forward with this simple definition, it will be easy to make everything consistent. If it means breaking changes, we will either need to do is quickly before v11 (going out soon) or it'll need to wait for v12.</blockquote>

I agree with that meaning of an absolute path, of course (who could argue otherwise) but I would _love_ it if you would support "absolute" paths as relative to product root, as worked in past versions.  Maybe a config option?  Or am I the only one who would ever use it?

If you decide to do that and can do it for v11, great!  If you decide to do it for v12 and go with true absolute for v11, that will break things for us.  Please let me know and we'll wait for v12.

If you decide not to do it but make everything consistent with true absolute, then when we upgrade we'll bite the bullet and update the paths in our files.

And then everything will just work... :)

**Reply by lmiller1990** [reply]
I don't think a config makes sense - better to add something new. It's hard to make a case for just a handful of uses cases, though. I wonder if we can write some code that'll just do something like

```js
function absoluteFromProjectRoot (p: string) {
  return path.join(project_root, p)
}
```

You can probably infer the project root by traversing up until you hit a `.git` directory, or something.

**Reply by grivard** [reply]
I certainly agree, if I'm the only one using "absolute" references to project root it wouldn't make sense to implement something special for that.  We can live with full absolute paths working consistently everywhere.  Worst case our import/require/include spells out the full path, or maybe we can do something like `const root = '/full/path/'` along with `import root + 'path/from/root/file.js'`.

Again, I appreciate your attention to this.  We have a workaround in place now (configs at project root) and will await a version where we can move them back into the config folder.  Thanks!
